ESPS/general/man/man3/atoarray.3
.\" Copyright (c) 1986-1990 Entropic Speech, Inc.
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)atoarray.3	1.6 09 May 1997 ESI/ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH ATOARRAY 3\-ESPSu 09 May 1997
.SH NAME
.nf
atoarrayf \- convert ASCII file to data array of floats
atoarray \- convert ASCII file to data array of specified numeric type
atoarrays \- convert ASCII file to data array of strings
.fi
.SH SYNOPSIS
.nf
.ft B
#include <esps/esps.h>

float *
atoarrayf(strm, nvalues, max)
FILE *strm;
long *nvalues;
float *max;

char *
atoarray(strm, type, nvalues, max)
FILE *strm;
int type;
long *nvalues;
double *max;

char **
atoarrays(strm, nvalues, max)
FILE    *strm;	
long    *nvalues;
int      *max;	

.ft
.fi
.SH DESCRIPTION
.I atoarrayf
and
.I atoarray
read the Ascii file stream
.I strm,
which is assumed to contain a series of numerical values
separated by white space or new lines;
they convert the values to the required type
and return a pointer to the converted data.
.PP
The required data type is FLOAT for
.I atoarrayf.
For
.I atoarray
it is indicated by the value of
.I type,
which must be one of the integer constants
DOUBLE, FLOAT, LONG, SHORT, BYTE,
DOUBLE_CPLX, FLOAT_CPLX, LONG_CPLX, SHORT_CPLX, and BYTE_CPLX
defined in
.I esps/esps.h.
See the table in the
.RI FEA (5-ESPS)
man page; non-numeric types are not allowed.
If the type is complex, the input values are taken in pairs;
the first value in a pair gives the real part of a complex number,
and the second value gives the imaginary part.
An input value is written as an optionally signed string of decimal digits,
which, for types DOUBLE, FLOAT, DOUBLE_CPLX, and FLOAT_CPLX,
may contain a decimal point and may be followed by an
.I E
or
.I e
and an optionally signed integer.
.PP
\fIatoarrays\fP reads the Ascii file stream \fIstrm\fP and returns an
array of strings, one string per line in file stream.  Trailing line
feeds are dropped before each line is added to the string array.  
.PP
For all three functions, the number of values read is returned through
the parameter
.I nvalues.
(For complex types, that is the number of complex values,
not the number of real values.)
.PP
The maximum absolute value or maximum string length (in the case of 
\fIatoarrays\fP) encountered is returned through
.I max
if
.I max
is non-NULL.
A NULL value for
.I max
is treated as an error by
.I atoarrayf.
A value of NULL for
.I max
is a signal to
.I atoarray
and
.I atoarrays
that the maximum absolute value or maximum string length is not wanted.
.PP
The returned values from 
.I atoarrayf
and 
.I atoarray
is a pointer to the first byte of the storage area for the results.
The storage is allocated by
.IR malloc (3c)
and may be freed by
.IR free (3c).
Before being used, the pointer returned by
.I atoarray
must be cast to a pointer type depending on
.I type
.RI ( "e.g."
(double *) for DOUBLE, (byte_cplx *) for BYTE_CPLX).
The value returned by
.I atoarrayf
does not require a cast.
.PP
.I atoarrayf,
.I atoarray
and
.I atoarrays
can be used on standard input.  
.SH EXAMPLE
.nf
    float   *sdata;	 
    float   maxvalf;
    double  maxvald;
    long    points;
    FILE    *instrm;
     .
     .
     .
    sdata = atoarrayf(instrm, &points, &maxvalf);
    	/* or */
    sdata = (float *) atoarray(instrm, FLOAT, &points, &maxvald);
    /*
     * process data in sdata
     */
.fi
.SH DIAGNOSTICS
The functions will an assertion failure and a diagnostic message if any of
the following are true:
.PP
The function is unable to allocate sufficient memory.
.PP
A NULL pointer is passed for  
.I strm
or
.I nvalues.
.PP
A NULL pointer is passed to
.I atoarrayf
for
.I max.
.PP
An illegal value is passed to
.I atoarray
for
.I type.
.PP
The function encounters an illegal format in the input.
.PP
.I atoarray
finds an odd number of real input values while reading complex data.
.SH BUGS
None known.
.SH "FUTURE CHANGES"
Version of
.I atoarray
that will read into a user-supplied data area
and will stop reading after a specified maximum number of elements.
.SH SEE ALSO
\fIaddstr\fP (3\-\s-1ESPS\s+1), malloc(3c), free(3c).
.SH AUTHOR
.I atoarrayf
and
.I atoarrays
by John Shore.
.I atoarray
by Rodney Johnson, based on
.I atoarrayf.


ESPS/general/src/lib/atoarray.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  John Shore (atoarrayf, atoarray2), 
 *              Rodney Johnson (atoarray)
 * Checked by:  Alan Parker (atoarrayf)
 * Revised by:
 *
 * Brief descriptions: 
 *
 *     atoarrayf, converts an ASCII file to a float array		
 *     atoarrays, converts an ASCII file to a string array	
 *     atoarray,  which converts an ASCII file to an array of any type.	
 *
 * The file contains numerical values separated by white space or new  
 * lines.							       
 */

static char *sccs_id = "@(#)atoarray.c	1.9	2/20/96	ERL/ESI";


/*
 * include files
 */

#include <stdio.h>
#include <esps/esps.h>
#if !defined(LINUX_OR_MAC)
#include <esps/unix.h>
#endif
#include <math.h>

/*
 * defines
 */

#define DATA_CHUNK 25000	/* number of ASCII points handled at a time */
#define ABS(a) (((a) >= 0) ? (a) : -(a))

/*
 * system functions and variables
 */

#if !defined(DEC_ALPHA) && !defined(HP700) && !defined(LINUX_OR_MAC)
char	    *realloc();
char	    *calloc();
double	    sqrt();
void        addstr();
#endif
/*
 * ESPS functions
 */
  
int 	    typesiz();
char        *savestring();

/*
 * local functions
 */

static int  numeric();

/*
 * main functions
 */

float *
atoarrayf(strm, nvalues, max)
    FILE    *strm;	/*ASCII file stream*/
    long    *nvalues;	/*pointer to number of values found*/
    float   *max;	/*pointer to return maximum absolute value found*/
{
    float *data;	    /*pointer to data*/
    unsigned dsize = 0;		/* count of currently allocated floats */
    unsigned cnt = 0;		/* count within current chunk */
    int nchar;			/* number of characters matched by fscanf */
    spsassert(strm != NULL, "atoarrayf: NULL file pointer");
    spsassert(nvalues != NULL, "atoarrayf: NULL pointer for nvalues");
    spsassert(max != NULL, "atoarrayf: NULL pointer for max value");
    *max = 0.0;
    *nvalues = 0;	
    /*
     * allocate space for first DATA_CHUNK of floats
     */
    data = (float*) calloc ((unsigned) DATA_CHUNK, sizeof(float));
    dsize = DATA_CHUNK;
    spsassert(data != NULL, "memory allocation failure");
    while ((nchar = fscanf(strm, "%f", &data[*nvalues])) != EOF) {
	spsassert(nchar != 0, "atoarrayf: input not in float format");
	if (ABS(data[*nvalues]) > *max) *max = ABS(data[*nvalues]);
	(*nvalues)++;
	cnt++;
	/*
	 * get more space if we've used up the chunk
	 */
	 if (cnt == DATA_CHUNK) {
	    dsize += DATA_CHUNK;
	    data = (float*) realloc
	        ((char *)data, (unsigned)dsize*sizeof(float));
	    spsassert(data != NULL, "atoarrayf: memory allocation failure");
	    cnt = 0;
	 }
    }
    /*
     * don't return more memory than needed
     */
    data = (float*) realloc((char *)data, (unsigned) *nvalues * sizeof(float));
    spsassert(data != NULL, "atoarrayf: realloc failed to reduce Size");
    return data;
}


#define MAXLINE 250
char **
atoarrays(strm, nvalues, max)
    FILE    *strm;	/*ASCII file stream*/
    long    *nvalues;	/*pointer to number of values found*/
    int      *max;	/*pointer to return maximum string length found*/
{
    char **data = NULL;	    /*pointer to data*/
    char *string_item = (char *) malloc(MAXLINE);
    int length = 0;		/* length of latest string */

    spsassert(strm != NULL, "atoarrays: NULL file pointer");
    spsassert(nvalues != NULL, "atoarrays: NULL pointer for nvalues");

    if (max != NULL) 
      *max = 0;

    *nvalues = 0;	

    while (fgets(string_item, MAXLINE, strm) != NULL) {

	length = strlen(string_item);

	if (length > 0) {

	    /* remove trailing line feed*/

	    if (string_item[length-1] == '\n')
	      string_item[length-1] = '\0';

	    length = strlen(string_item);

	    if (length > 0) {

		addstr(savestring(string_item), &data);

		if ((max != NULL) && (length > *max))
		  *max = length; 

		(*nvalues)++;
	      }
	  }
      }
    return data;
}


#define CHUNK_SIZE  (DATA_CHUNK * sizeof(float))

char *
atoarray(strm, type, nvalues, max)
    FILE    *strm;
    int	    type;
    long    *nvalues;
    double  *max;
{
    char	*data;		/* pointer to data*/
    unsigned	dsize;		/* count of currently allocated elements */
    unsigned	d_chunk;	/* number of elements to allocate at a time */
    unsigned	tsize;		/* size of one data element */
    long	cnt;		/* count of items read */
    int	    	nitem;		/* number of items matched by fscanf */
    double  	absval;		/* absolute value of data element */
    double  	reval, imval;	/* real & imaginary parts of data element */
    double  	maxabs;		/* max absolute value */

    spsassert(strm, "atoarray: NULL file pointer");
    spsassert(nvalues, "atoarray: NULL pointer for nvalues");
    spsassert(numeric(type), "atoarray: non-numeric type code");

    tsize = typesiz(type);
    cnt = 0;
    if (max) maxabs = 0.0;

    /*
     * allocate space for first data chunk
     */

    d_chunk = CHUNK_SIZE / tsize;
    data = calloc(d_chunk, tsize);
    spsassert(data, "atoarray: can't allocate space for data");
    dsize = d_chunk;
    
    for ( ; ; )		/* Will break out of loop on EOF or bad data. */
    {
	switch(type)
	{
	case DOUBLE: {
	    double	*arr = (double *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%lf", &arr[cnt])) == 1);
		 cnt++)
	    {
		if (max)
		{
		    absval = arr[cnt];
		    if (absval < 0) absval = -absval;
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case FLOAT: {
	    float	*arr = (float *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%f", &arr[cnt])) == 1);
		 cnt++)
	    {
		if (max)
		{
		    absval = arr[cnt];
		    if (absval < 0) absval = -absval;
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case LONG: {
	    long	*arr = (long *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%ld", &arr[cnt])) == 1);
		 cnt++)
	    {
		if (max)
		{
		    absval = arr[cnt];
		    if (absval < 0) absval = -absval;
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case SHORT: {
	    short	*arr = (short *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%hd", &arr[cnt])) == 1);
		 cnt++)
	    {
		if (max)
		{
		    absval = arr[cnt];
		    if (absval < 0) absval = -absval;
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case BYTE: {
	    char	*arr = (char *) data;
	    short	item;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%hd", &item)) == 1);
		 cnt++)
	    {
		arr[cnt] = item;
		if (max)
		{
		    absval = arr[cnt];
		    if (absval < 0) absval = -absval;
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case DOUBLE_CPLX: {
	    double_cplx	*arr = (double_cplx *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%lf%lf",
				     &arr[cnt].real, &arr[cnt].imag)) == 2);
		 cnt++)
	    {
		if (max)
		{
		    reval = arr[cnt].real;
		    if (reval < 0) reval = -reval;
		    imval = arr[cnt].imag;
		    if (imval < 0) imval = -imval;

		    if (reval > imval)
		    {
			imval /= reval;
			absval = reval * sqrt(1.0 + imval*imval);
		    }
		    else if (imval > reval)
		    {
			reval /= imval;
			absval = imval * sqrt(1.0 + reval*reval);
		    }
		    else absval = reval * sqrt(2.0);

		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case FLOAT_CPLX: {
	    float_cplx	*arr = (float_cplx *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%f%f",
				     &arr[cnt].real, &arr[cnt].imag)) == 2);
		 cnt++)
	    {
		if (max)
		{
		    reval = arr[cnt].real;
		    if (reval < 0) reval = -reval;
		    imval = arr[cnt].imag;
		    if (imval < 0) imval = -imval;

		    if (reval > imval)
		    {
			imval /= reval;
			absval = reval * sqrt(1.0 + imval*imval);
		    }
		    else if (imval > reval)
		    {
			reval /= imval;
			absval = imval * sqrt(1.0 + reval*reval);
		    }
		    else absval = reval * sqrt(2.0);
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case LONG_CPLX: {
	    long_cplx	*arr = (long_cplx *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%ld%ld",
				     &arr[cnt].real, &arr[cnt].imag)) == 2);
		 cnt++)
	    {
		if (max)
		{
		    reval = arr[cnt].real;
		    imval = arr[cnt].imag;
		    absval = sqrt(reval*reval + imval*imval);
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case SHORT_CPLX: {
	    short_cplx	*arr = (short_cplx *) data;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%hd%hd",
				     &arr[cnt].real, &arr[cnt].imag)) == 2);
		 cnt++)
	    {
		if (max)
		{
		    reval = arr[cnt].real;
		    imval = arr[cnt].imag;
		    absval = sqrt(reval*reval + imval*imval);
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	case BYTE_CPLX: {
	    byte_cplx	*arr = (byte_cplx *) data;
	    short	re_item, im_item;

	    for ( ;
		 (cnt < dsize
		  && (nitem = fscanf(strm, "%hd%hd",
				     &re_item, &im_item)) == 2);
		 cnt++)
	    {
		arr[cnt].real = re_item;
		arr[cnt].imag = im_item;
		if (max)
		{
		    reval = arr[cnt].real;
		    imval = arr[cnt].imag;
		    absval = sqrt(reval*reval + imval*imval);
		    if (absval > maxabs) maxabs = absval;
		}
	    }
	    break; }
	}
	if (cnt != dsize) break; /* Exit loop on EOF or bad data. */
	else
	{
	    dsize += d_chunk;
	    data = realloc(data, dsize*tsize);
	    spsassert(data, "atoarray: can't reallocate space for data");
	}
    }
    switch (nitem)
    {
    case 0:
	Fprintf(stderr, "%s: badly formatted input item.\n", "atoarray");
	exit(1);
	break;
    case 1:
	Fprintf(stderr,
		"%s:  badly formatted input item or missing re/im part.\n",
		"atoarray");
	exit(1);
	break;
    case EOF:
	data = realloc(data, dsize*tsize);
	spsassert(data, "atoarray: realloc failed to reduce size");
	if (max) *max = maxabs;
	*nvalues = cnt;
	return data;
	break;
    default:
	Fprintf(stderr, "%s: error reading data. fscanf returned %d\n",
	       "atoarray", nitem);
	exit(1);
	break;
    }

    /*NOTREACHED*/
}



/*
 * Check for numeric data type.
 */

static int
numeric(type)
    int	    type;
{
    switch (type)
    {
    case BYTE:
    case SHORT:
    case LONG:
    case FLOAT:
    case DOUBLE:
    case BYTE_CPLX:
    case SHORT_CPLX:
    case LONG_CPLX:
    case FLOAT_CPLX:
    case DOUBLE_CPLX:
	return YES;
	break;
    default:
	return NO;
	break;
    }

    /*NOTREACHED*/
}

